Adds TopicModel.infer methods

sagemintblue opened the pull request:
The infer methods iteratively call trainDocTopicModel while keeping track of relative document perplexity difference between iterations. This simplifies iteration till convergence.

Also:
- Updates CVB0DocInferenceMapper to use infer method.
- More cleanup of CVBConfig and CVB0Driver, including additional params to support use of infer method.

sagemintblue committed:
Adds TopicModel.infer methods

The infer methods iteratively call trainDocTopicModel while keeping track of
relative document perplexity difference between iterations. This simplifies
iteration till convergence.

Also:

- Updates CVB0DocInferenceMapper to use infer method.

- More cleanup of CVBConfig and CVB0Driver, including additional params to
  support use of infer method.

sagemintblue commented:
Forgot to mention in that commit message; Critical issue fixed in CVB0Driver regarding use of `FileSystem.get(...)`:

Basically, I was testing the code, and noticed that the driver was reporting some paths as non-existent when they clearly did exist. The use of

http://hadoop.apache.org/common/docs/r0.20.2/api/org/apache/hadoop/fs/FileSystem.html#get(java.net.URI,%20org.apache.hadoop.conf.Configuration)

was causing problems. I've replaced all instances of this with

http://hadoop.apache.org/common/docs/r0.20.2/api/org/apache/hadoop/fs/FileSystem.html#get(org.apache.hadoop.conf.Configuration)

jakemannix commented on core/src/main/java/org/apache/mahout/clustering/lda/cvb/CVBConfig.java:
```diff
@@ -62,10 +67,16 @@
   public static final int NUM_REDUCE_TASKS_DEFAULT = 1;
   public static final float MODEL_WEIGHT_DEFAULT = 1f;
-  
+  public static final float MIN_RELATIVE_PERPLEXITY_DIFF_DEFAULT = 1e-8f;
```
It's possible that this magic convergence number may need to be modified or something.  It looks good on our data, but... :)

jakemannix commented on core/src/main/java/org/apache/mahout/clustering/lda/cvb/CVBConfig.java:
```diff
@@ -62,10 +67,16 @@
   public static final float MODEL_WEIGHT_DEFAULT = 1f;
-  
+  public static final float MIN_RELATIVE_PERPLEXITY_DIFF_DEFAULT = 1e-8f;
+  public static final int MAX_INFERENCE_ITERATIONS_PER_DOC_DEFAULT = 100;
```
This may be a bit high for the general use case as well.   Possibly us overriding it to be 100 is the right thing, but maybe 20 is more reasonable in general?  Not sure.  Maybe just leave a comment about it.

jakemannix commented on core/src/main/java/org/apache/mahout/clustering/lda/cvb/CVBConfig.java:
```diff
@@ -294,7 +309,25 @@ public CVBConfig setModelWeight(float modelWeight) {
+    return maxInferenceItersPerDoc;
+  }
+
+  public CVBConfig setMaxInferenceItersPerDoc(int maxInferenceItersPerDoc) {
```
We should probably start using some good Preconditions checking here - TODO for later - make sure negative values aren't put in here, etc...

jakemannix replied:
Nevermind, noticed the check() method below, good work!

sagemintblue committed:
Adds max_inf_doc_topic_iters cmdline argument parsing

sagemintblue committed:
Updates FileSystem.get(...) invocations to use URI param

For some reason this now works in my environment, and is necessary for model paths to be found. Go figure.

sagemintblue committed:
Fixes inference halting logic to prevent run-away iteration
